daemon/SecretsImpl/dataprotector.cpp

```diff
@@ -0,0 +1,270 @@
+        bool cantopen = file->open(QIODevice::ReadOnly);
+        if (!cantopen) {
+            // Loss of any individual file here doesn't yet necessarily mean an error
+            qWarning(lcSailfishSecretsDaemon) << "can't open file:" << file->fileName();
+        }
+        return cantopen;
+    });
+
+    // If none of the files could be opened, we are very unlucky
+    if (allFilesLost) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not open any files. Data is irretrieveable.";
+        m_hasError = true;
+        return QByteArray();
+    }
+
+    // Read all files
+    QVector<QByteArray> fileContents;
+    std::transform(files.begin(), files.end(), std::back_inserter(fileContents), [](QSharedPointer<QFile> &file) {
+        return file->readAll();
+    });
+
+    std::remove_if(fileContents.begin(), fileContents.end(), [](QByteArray &byteArray) {
+        return byteArray.isEmpty();
+    });
+
+    //  If none of the files could be read, again we are unlucky
+    if (fileContents.size() == 0) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not read any of the files. Data is irretrievable.";
+        m_hasError = true;
+        return QByteArray();
+    }
+
+    // Assuming that any of the byte arrays can differ from any of the others,
+    // count how many times each occours. Ideally we end up with 3 occourences
+    // on the first index, and 0 on the others.
+    QVector<int> occourences(fileContents.size());
+    for (const QByteArray &currentData : fileContents) {
+        for (int i = 0; i < fileContents.size(); i++) {
+            if (currentData == fileContents.at(i)) {
+                occourences[i]++;
+                break;
+            }
+        }
+    }
+
+    // Assume that the version that occours the most is the correct version.
+    int maxOccourenceIndex = 0;
+    for (int i = 0; i < occourences.size(); i++) {
+        if (occourences[i] > occourences[maxOccourenceIndex]) {
+            maxOccourenceIndex = i;
+        }
+    }
+    int minimumValidOccourences = fileContents.size() / 2 + 1;
+    if (occourences[maxOccourenceIndex] < minimumValidOccourences) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not decide which file is correct, there is no majority.";
+        m_hasError = true;
+        return QByteArray();
+    }
+
+    m_data = fileContents[maxOccourenceIndex];
+    return m_data;
+}
+
+void DataProtector::setData(const QByteArray &bytes)
+{
+    QDir dir(m_path);
+
+    // Check if the main directory exists, create if not
+    if (!dir.exists()) {
+        if (!dir.mkpath(m_path)) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't create protected root directory:" << m_path;
+            m_hasError = true;
+            return;
+        }
+
+        qCDebug(lcSailfishSecretsDaemon) << "Protected root directory didn't exist, so the data is empty.";
+    }
+
+    auto oldDirectoryEntries = dir.entryInfoList(QDir::Dirs);
```

Comment:
autos are not helping here more like hindering legibility

Reply:
That's their purpose isn't it?

Reply:
Sorry.  
I know you guys hate `auto` and will remove them when I finalize this PR.

For me personally, it actually improves legibility.

Reply:
Your own code is always readable for yourself, at least right after you wrote it. There's nothing surprising about that ))